### DISCUSSION TITLE
set_session cannot be used inside a transaction with pool_pre_ping

### DISCUSSION THREAD

**Discussion by arpekkar** [sqlalchemy/sqlalchemy#8009] [Usage Questions]
### Describe the bug

Because connection drop issues I needed to start using pool_pre_ping with sqlalchemy. Im using psycopg2 as an adapter

Now I'm facing a new issue:

`sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) set_session cannot be used inside a transaction`

### To Reproduce

```python
engines = {
      "writer": create_engine(
          f"postgresql+psycopg2://{username}:{password}@{writer_endpoint}:{port}/{database}",
          logging_name="writer",
          pool_pre_ping=True,
      )
}
```

I'm opening a session with context manager and commiting the transaction explicitly:

```
    @contextmanager
    def session_scope(self, session_cls, **kwargs):
        session = session_cls(**kwargs)
        try:
            yield session
            session.commit()
        except Exception:
            session.rollback()
            raise
        else:
            session.close()
```

```
with self.session_scope(self.session) as session:
    # Doing my query here

```

### Error

```
File "/opt/python/sqlalchemy/orm/query.py", line 2709, in all
return self._iter().all()
File "/opt/python/sqlalchemy/orm/query.py", line 2844, in _iter
result = self.session.execute(
File "/opt/python/sqlalchemy/orm/session.py", line 1688, in execute
conn = self._connection_for_bind(bind)
File "/opt/python/sqlalchemy/orm/session.py", line 1529, in _connection_for_bind
return self._transaction._connection_for_bind(
File "/opt/python/sqlalchemy/orm/session.py", line 747, in _connection_for_bind
conn = bind.connect()
File "/opt/python/sqlalchemy/engine/base.py", line 3165, in connect
return self._connection_cls(self, close_with_result=close_with_result)
File "/opt/python/sqlalchemy/engine/base.py", line 96, in __init__
else engine.raw_connection()
File "/opt/python/sqlalchemy/engine/base.py", line 3244, in raw_connection
return self._wrap_pool_connect(self.pool.connect, _connection)
File "/opt/python/sqlalchemy/engine/base.py", line 3214, in _wrap_pool_connect
Connection._handle_dbapi_exception_noconnection(
File "/opt/python/sqlalchemy/engine/base.py", line 2068, in _handle_dbapi_exception_noconnection
util.raise_(
File "/opt/python/sqlalchemy/util/compat.py", line 207, in raise_
raise exception
File "/opt/python/sqlalchemy/engine/base.py", line 3211, in _wrap_pool_connect
return fn()
File "/opt/python/sqlalchemy/pool/base.py", line 307, in connect
return _ConnectionFairy._checkout(self)
File "/opt/python/sqlalchemy/pool/base.py", line 800, in _checkout
result = pool._dialect.do_ping(fairy.connection)
File "/opt/python/sqlalchemy/dialects/postgresql/psycopg2.py", line 850, in do_ping
dbapi_connection.autocommit = True
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) set_session cannot be used inside a transaction
(Background on this error at: http://sqlalche.me/e/14/f405)
``` 

### Versions

- OS: Linux
- Python: 3.8.10
- SQLAlchemy: 1.4.20
- Database: PostgreSQL 13.3
- DBAPI: Psycopg 2.9.1

**Comment by zzzeek** [comment]
some improvements have been made to psycopg2 do_ping as part of #7930 which may resolve the problem you're having (that is, always upgrade to the latest 1.4.x version, 1.4.36 right now, before reporting any issues).  otherwise no issue can be reproduced so you want to be careful about any custom "on connect" event handlers and such.

**Comment by harpener** [comment]
@zzzeek Hi, I am having a similar problem with my project and I see the same exception with `SQLAlchemy[postgresql]==1.4.36` and `psycopg2==2.9.3`.

For the record, I am using these flags on `sessionmaker`:
```
autocommit=False
autoflush=False
expire_on_commit=False
```

**Reply by zzzeek** [reply]
Here's a complete test.  can you modify it to show your error?  thanks


```py
from contextlib import contextmanager

from sqlalchemy import create_engine
from sqlalchemy import text
from sqlalchemy.orm import Session

engine = create_engine(
    "postgresql+psycopg2://scott:tiger@localhost/test",
    logging_name="writer",
    pool_pre_ping=True,
    echo=True
)


@contextmanager
def session_scope(session_cls, **kwargs):
    session = session_cls(**kwargs)
    try:
        yield session
        session.commit()
    except Exception:
        session.rollback()
        raise
    else:
        session.close()


for i in range(10):
    with session_scope(
        Session,
        bind=engine,
        autocommit=False,
        autoflush=False,
        expire_on_commit=False,
    ) as session:
        session.execute(text("select 1")).all()

```

**Reply by harpener** [reply]
Thanks for the example.

I should also mention that I use `scoped_session` for scoping to process and thread. And the transaction is essentially a class that returns a current session via `scoped_session()` on enter and commits/rollbacks the session (after obtaining it the same way again) on exit as well.

The important thing is that the exception is raised only rarely and I have not been able to fully understand why it happens. If I could reproduce it, it would surely help, but not even disconnecting the database between two queries within one transaction does not seem to be the problem here.

**Reply by harpener** [reply]
One other small point, it started happening only after I upgraded SQLA 1.3.x to 1.4.36.

**Reply by zzzeek** [reply]
using multiple threads ?    does the error go away when you set threadcount to 1 ?

**Reply by harpener** [reply]
That's actually a very good question, I'm using `concurrent.futures.ThreadPoolExecutor` near the affected code and now that I am thinking about it, the database operations might not be thread-safe at that moment, because the scoping happens much sooner. 🤔

**Reply by harpener** [reply]
I can confirm now that the error disappeared after disabling threads. As predicted, the session is not properly scoped to on-the-fly thread.
